### ISSUE TITLE
Failure in Fetching Global Sampling Rules

### ISSUE THREAD

**Issue by Edward-Francis** [aws/aws-xray-sdk-python#162]
I'm having problems with the `aws_xray_sdk` package fetching global sampling configuration. If I understand correctly when the recorder is initialised the Sampler attempts to fetch rules from the `RulePoller()` and then re-fetches these rules every 5 minutes period. However the following test code doesn't seem to fetch the configuration unless I specifically force it too. This setup is running in Docker images and is using the [X-Ray daemon (v3.0.1)](https://hub.docker.com/r/amazon/aws-xray-daemon/).

## CODE

```
import logging
from aws_xray_sdk.core import xray_recorder

logging.basicConfig(level=logging.DEBUG)
xray_recorder.configure(service="My Application")

segment = xray_recorder.begin_segment("Test Segment")
segment.put_annotation("test", "1")
xray_recorder.end_segment()

# Un-comment forces the sampler to gather global rules 
# xray_recorder.sampler._rule_poller._refresh_cache()
# print(xray_recorder.sampler._cache.rules) # [<SamplingRule>, <SamplingRule>]

print(xray_recorder.sampler._cache.rules) # []
```

## DEBUG

Without force `_refresh_cache()` - missing *DEBUG:urllib3.connectionpool:http Response*:
```
DEBUG:botocore.hooks:Event request-created.xray.GetSamplingRules: calling handler <bound method RequestSigner.handler of <botocore.signers.RequestSigner object at 0x7f6034a4cd68>>
DEBUG:botocore.hooks:Event choose-signer.xray.GetSamplingRules: calling handler <function set_operation_specific_signer at 0x7f6035200488>
DEBUG:botocore.endpoint:Sending http request: <AWSPreparedRequest stream_output=False, method=POST, url=http://xray-daemon:2000/GetSamplingRules, headers={'User-Agent': b'Botocore/1.12.59 Python/3.6.6 Linux/4.14.106-79.86.amzn1.x86_64', 'Content-Length': '0'}>
DEBUG:urllib3.util.retry:Converted retries value: False -> Retry(total=False, connect=None, read=None, redirect=0, status=None)
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): xray-daemon:2000
INFO:aws_xray_sdk.core.sampling.sampler:No effective centralized sampling rule match. Fallback to local rules.
```

With force `_refresh_cache()`:
```
DEBUG:botocore.hooks:Event request-created.xray.GetSamplingRules: calling handler <bound method RequestSigner.handler of <botocore.signers.RequestSigner object at 0x7f25c5774da0>>
DEBUG:botocore.hooks:Event choose-signer.xray.GetSamplingRules: calling handler <function set_operation_specific_signer at 0x7f25c5f29488>
DEBUG:botocore.endpoint:Sending http request: <AWSPreparedRequest stream_output=False, method=POST, url=http://xray-daemon:2000/GetSamplingRules, headers={'User-Agent': b'Botocore/1.12.59 Python/3.6.6 Linux/4.14.106-79.86.amzn1.x86_64', 'Content-Length': '0'}>
DEBUG:urllib3.util.retry:Converted retries value: False -> Retry(total=False, connect=None, read=None, redirect=0, status=None)
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): xray-daemon:2000
DEBUG:urllib3.connectionpool:http://xray-daemon:2000 "POST /GetSamplingRules HTTP/1.1" 200 725
DEBUG:botocore.parsers:Response headers: {'Content-Length': '725', 'Content-Type': 'application/json', 'Date': 'Fri, 07 Jun 2019 10:34:30 GMT', 'X-Amzn-Requestid': 'd508d0a2-890f-11e9-96dd-5ff356ea9bca'}
```
Any help would be appreciated!

**Comment by chanchiem** [comment]
Hey there!

Thanks for reaching out to us about the issue you're having. I have a feeling that the issue you're having is due to the fact that the rule poller threads haven't been started yet, which is why manually calling the helper function `_refresh_cache()` acquires the sampling rules. When you first generate your segment and attempt to get a sampling decision, you would usually call xray_recorder.sampler.should_trace() for a given sampling request. And when this happens, if the pollers have not been started yet, it would then designated it to acquire sampling rules. [Here](https://github.com/aws/aws-xray-sdk-python/blob/master/aws_xray_sdk/core/sampling/sampler.py#L62) is the logic that handles that.

This is actually by design. For X-Ray instrumented applications that don't necessarily need sampling rules, this prevented high volume traffic from being pushed to the service with GetSamplingRules calls when the target application doesn't need the sampling decision.

So in short, what I would recommend doing is either calling should_trace from the sampler, or manually call the start() method in the sampler.

Please let me know if this helps!

**Comment by Edward-Francis** [comment]
Hi,

Thanks for your help - yes you were right the sampler and rule poller hadn't been started yet - so couldn't fetch the rule configuration. I opted for using `xray_recorder.sampler.start()` before starting creating the segment.

Thanks again!
